### ISSUE TITLE
Multiple unsecure HTTP calls

### ISSUE THREAD

**Issue by zachaysan** [crystal-lang/crystal#1922]
On [this](http://crystal-lang.org/docs/installation/on_debian_and_ubuntu.html) page and others, as well as in the vagrant file and I believe the deb as well (shudder) do not have HTTPS endpoints.

`curl http://dist.crystal-lang.org/apt/setup.sh | sudo bash`

Is only safe if there is an `s` after the `http`. Otherwise I'm letting the guy that provides the wifi at the cafe I'm sitting at access to my system. If it is a matter of affording the certificate, I'll pay for it myself. Crystal is too far along to have these types of vulnerabilities.

**Comment by zachaysan** [comment]
I figured I should actually get involved and fix it, or put up the money if it bothers me so much, so I made a bounty for it:

https://www.bountysource.com/issues/28722212-multiple-unsecure-http-calls

Sorry for coming off as dickish earlier.

**Comment by EvanHahn** [comment]
Is this made easier by [Let's Encrypt](https://letsencrypt.org/)?

**Comment by zachaysan** [comment]
@EvanHahn Yes. Not just easier, but cheaper too.

**Comment by ozra** [comment]
There are plenty of free certificate options, so financials shouldn't be a problem.

**Comment by jhass** [comment]
So this landed in my triage today, it's a bit sad to get ignored for so long. Given this is currently hosted on Amazon S3, currently I see the following solutions here:

* Give up the custom domain, Amazon ships wildcard certificates for `*.s3.amazonaws.com`, accessing a bucket via `https://s3.amazonaws.com/bucket_name` is possible too.
* Enable CloudFront and point the custom domain at it, then use the [SNI cert option](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/SecureConnections.html#cnames-and-https-switch-dedicated-to-sni). CloudFront will incur additional costs and also additional costs for HTTPS connections.
* Go through CloudFlare, which provides certificates from their own CA for traffic going through them. Care should be taken to use a secure connection to the S3 bucket via the first option.
* Get a cheap/sponsored VPS and setup a reverse proxy. This is basically homebrew CloudFlare, so the same note regarding a secure connection to S3 applies.
* Get a cheap/sponsored VPS/shared hosting plan and set it up as a file host. In the VPS case certificate issuance can be automated using LetsEncrypt, in the shared hosting plan automated certificate handling should be included.
* Research services similar to S3 which include easy TLS/HTTPS setup at no additional costs.

For the options that require to obtain a custom cert, StartSSL and WoSign provide free certificates with lifetimes longer than the 90 days provided by LetsEncrypt.

**Comment by waj** [comment]
Sorry for taking so long to fix this. I just enabled CloudFront with a certificate issued (free) by AWS.
HTTP requests are still enabled for compatibility but we could deprecate them in the future once most people migrated their repository configurations. Official announcement coming soon to advise about that.

**Comment by jhass** [comment]
I would expect apt and yum/zapper/dnf to follow redirects, isn't it possible to just redirect to https?

**Comment by zachaysan** [comment]
Ultimately I'd expect all of crystal-lang.org to be HTTPS and the domain added to HSTS lists in order to stop a MITM from altering the installation instructions - but for now leaving the HTTP route or redirecting to HTTPS is _ok_. Perhaps a deprecation notice could be issued if a potential installer has attempted to use the non-HTTPS route?

**Comment by splattael** [comment]
After the commit e1d2fcd I am unable to build a Docker image anymore with:
`docker build -f Dockerfile.release .`

I get this error suggesting to install missing `apt-transport-https`.

```
Sending build context to Docker daemon 6.948 MB
Step 1 : FROM ubuntu:14.04
14.04: Pulling from library/ubuntu
56eb14001ceb: Pull complete 
7ff49c327d83: Pull complete 
6e532f87f96d: Pull complete 
3ce63537e70c: Pull complete 
Digest: sha256:d67ef8e385f1c8b13d8c3e7622dc31b51d07e5623c1d034ebe2acb14a11fb30d
Status: Downloaded newer image for ubuntu:14.04
 ---> 38c759202e30
Step 2 : RUN apt-key adv --keyserver keys.gnupg.net --recv-keys 09617FD37CC06B54 &&   echo "deb https://dist.crystal-lang.org/apt crystal main" > /etc/apt/sources.list.d/crystal.list &&   apt-get update &&   apt-get install -y crystal gcc pkg-config libssl-dev git &&   apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 ---> Running in 9041f5b42170
Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.bEacGa9FQU --no-auto-check-trustdb --trust-model always --keyring /etc/apt/trusted.gpg --primary-keyring /etc/apt/trusted.gpg --keyserver keys.gnupg.net --recv-keys 09617FD37CC06B54
gpg: requesting key 7CC06B54 from hkp server keys.gnupg.net
gpg: key 7CC06B54: public key "Crystal Distribution Signing <dist-sign@crystal-lang.org>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
E: The method driver /usr/lib/apt/methods/https could not be found.
The command '/bin/sh -c apt-key adv --keyserver keys.gnupg.net --recv-keys 09617FD37CC06B54 &&   echo "deb https://dist.crystal-lang.org/apt crystal main" > /etc/apt/sources.list.d/crystal.list &&   apt-get update &&   apt-get install -y crystal gcc pkg-config libssl-dev git &&   apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*' returned a non-zero code: 100
```

This patch works for me:

```diff
diff --git a/Dockerfile.release b/Dockerfile.release
index d69fa7f..c86beb8 100644
--- a/Dockerfile.release
+++ b/Dockerfile.release
@@ -2,6 +2,8 @@ FROM ubuntu:14.04
 
 RUN \
   apt-key adv --keyserver keys.gnupg.net --recv-keys 09617FD37CC06B54 && \
+  apt-get update && \
+  apt-get install -y apt-transport-https && \
   echo "deb https://dist.crystal-lang.org/apt crystal main" > /etc/apt/sources.list.d/crystal.list && \
   apt-get update && \
   apt-get install -y crystal gcc pkg-config libssl-dev git && \
```

Note that `apt-get update` have to be called twice (once just to install https support for apt).

Do you want a PR?

**Comment by splattael** [comment]
I've just noted that my issue above has already been fixed in 0.18.5 (14 hours) ;)

Sorry for the noise!
